File: tests/e2e/fs.test.ts
/**
 * File system E2E — `fs_*` core commands.
 *
 * 실행:
 *   ./tests/e2e/run-fs.sh
 */
import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import puppeteer, { type Browser, type Page } from "puppeteer-core";

let browser: Browser;
let page: Page;

const core = <T = any>(request: Record<string, unknown>): Promise<T> =>
  page.evaluate(
    (req) => (window as any).__suji__.core(JSON.stringify(req)),
    request as any,
  ) as Promise<T>;

const FIXTURE_BASE = `/tmp/suji-fs-e2e-${process.pid}`;
const FIXTURE_DIR = `${FIXTURE_BASE}/nested`;
const FIXTURE_FILE = `${FIXTURE_DIR}/한글 file.txt`;

async function shellRun(cmd: string, args: string[]): Promise<{ exitCode: number }> {
  const proc = Bun.spawn([cmd, ...args]);
  await proc.exited;
  return { exitCode: proc.exitCode ?? 0 };
}

beforeAll(async () => {
  browser = await puppeteer.connect({
    browserURL: "http://localhost:9222",
    protocolTimeout: 30000,
    defaultViewport: null,
  });
  const pages = await browser.pages();
  expect(pages.length).toBeGreaterThan(0);
  page = pages[0];
  page.setDefaultTimeout(30000);

  await shellRun("rm", ["-rf", FIXTURE_BASE]);
});

afterAll(async () => {
  await shellRun("rm", ["-rf", FIXTURE_BASE]);
  await browser?.disconnect();
});

describe("fs core commands", () => {
  test("mkdir/write/read/stat/readdir round-trip", async () => {
    const mk = await core<{ success: boolean }>({ cmd: "fs_mkdir", path: FIXTURE_DIR, recursive: true });
    expect(mk.success).toBe(true);

    const wr = await core<{ success: boolean }>({
      cmd: "fs_write_file",
      path: FIXTURE_FILE,
      text: "hello\n한글",
    });
    expect(wr.success).toBe(true);

    const rd = await core<{ success: boolean; text: string }>({ cmd: "fs_read_file", path: FIXTURE_FILE });
    expect(rd.success).toBe(true);
    expect(rd.text).toBe("hello\n한글");

    const st = await core<{ success: boolean; type: string; size: number; mtime: number }>({
      cmd: "fs_stat",
      path: FIXTURE_FILE,
    });
    expect(st.success).toBe(true);
    expect(st.type).toBe("file");
    expect(st.size).toBeGreaterThan(0);
    expect(st.mtime).toBeGreaterThan(0);

    const ls = await core<{ success: boolean; entries: Array<{ name: string; type: string }> }>({
      cmd: "fs_readdir",
      path: FIXTURE_DIR,
    });
    expect(ls.success).toBe(true);
    expect(ls.entries).toContainEqual({ name: "한글 file.txt", type: "file" });
  });

  test("invalid path returns structured error", async () => {
    const r = await core<{ success: boolean; error: string }>({
      cmd: "fs_read_file",
      path: `${FIXTURE_BASE}/missing.txt`,
    });
    expect(r.success).toBe(false);
    expect(r.error).toBe("read");
  });

  test("empty path is rejected", async () => {
    const r = await core<{ success: boolean; error: string }>({ cmd: "fs_stat", path: "" });
    expect(r.success).toBe(false);
    expect(r.error).toBe("path");
  });
});

File: sdks/suji-go/fs/fs_test.go
package fs

import (
	"encoding/json"
	"testing"
)

func TestBuildFsRequests(t *testing.T) {
	cases := []struct {
		name string
		req  string
		cmd  string
	}{
		{"read", buildReadFileRequest("/tmp/a.txt"), "fs_read_file"},
		{"write", buildWriteFileRequest("/tmp/a.txt", "hello\nworld"), "fs_write_file"},
		{"stat", buildStatRequest("/tmp/a.txt"), "fs_stat"},
		{"mkdir", buildMkdirRequest("/tmp/dir", true), "fs_mkdir"},
		{"readdir", buildReadDirRequest("/tmp/dir"), "fs_readdir"},
		{"rm", buildRmRequest("/tmp/x", true, true), "fs_rm"},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			var got map[string]any
			if err := json.Unmarshal([]byte(tc.req), &got); err != nil {
				t.Fatalf("request is not valid JSON: %v", err)
			}
			if got["cmd"] != tc.cmd {
				t.Fatalf("cmd = %v", got["cmd"])
			}
		})
	}
}

func TestBuildFsRequestsEscapesStrings(t *testing.T) {
	req := buildWriteFileRequest(`/tmp/한글 "a".txt`, "line1\nline2\\tail")
	var got map[string]any
	if err := json.Unmarshal([]byte(req), &got); err != nil {
		t.Fatalf("request is not valid JSON: %v", err)
	}
	if got["path"] != `/tmp/한글 "a".txt` {
		t.Fatalf("path = %q", got["path"])
	}
	if got["text"] != "line1\nline2\\tail" {
		t.Fatalf("text = %q", got["text"])
	}
}
